IR Capture Alternate and Reset Pin

DaneHalle opened the pull request:
IR Capture Alternate:
Added a new input for game_imagerec.py, capture_screen(). capture_screen() is an alternate to SAVE_FRAMES where the user can determine when to save a frame as opposed to constantly saving frames while the script is running. 

Reset Pin information
Added new script, game_simple_reset.py which has base Reset functionality for the Trinket M0s. Adds an input for forcing a reset and also resets the Trinket during the on_prepare() phase.

Commit:
ninswitch: reset pin added

If connected, the pin resets the Trinket during on_prepare phase

DaneHalle commented:
I have tried to get the  flake8 test and linters to pass but it doesn't seem happy with the class definitions even though I have changed them. 
Not totally sure what the black test and linter is breaking on so let me know if I need to do anything.

ollipal commented on games/ninswitch/game_imagerec.py:
```diff
@@ -69,6 +89,12 @@ async def on_init(self):
 
+        self.io.register_inputs(
+            {
+                "capture_frame": capture_screen(self.cap),
```
Move this with the rest of the registered inputs above

DaneHalle replied:
In my testing, I found that having it before the other image_rec_task that were there normally caused issues. I will look into it again and see if I can get that working.

ollipal commented on games/ninswitch/game_imagerec.py:
```diff
@@ -31,10 +33,28 @@
     ((201, 650), (2, 2, 4)),
 ]
 
+class capture_screen(Switch):
```
I think this could be more elegant by having this class just flipping a new global variable `save_frame` between True/False, and sharing the same cap with `image_rec_main`. (also renaming SAVE_FRAMES --> SAVE_ALL_FRAMES).

Then in `image_rec_main` I would update:
```python
if SAVE_ALL_FRAMES or save_frame:
    cv2.imwrite(f"{SAVE_DIR_PATH}/{i}.jpg", frame)
    logging.info(f"SAVED {i}.jpg")
```
What do you think, does it make sense?

ollipal replied:
Also PEP8 naming: capture_screen --> CaptureScreen
I don't think this is enforced by our pipelines yet, but soon it might :D

DaneHalle replied:
I tried to do something like this and was unable to get it to work. You might be able to get it to work by using `global`.

ollipal commented on games/ninswitch/game_simple_reset.py:
```diff
@@ -1,11 +1,32 @@
 from games.ninswitch.ns_switch import NSSwitch
 from games.ninswitch.ns_dpad_switch import NSDPadSwitch
 from games.ninswitch.ns_joystick import NSJoystick
+import RPi.GPIO as GPIO
```
So far we have used `pigpio` library (http://abyz.me.uk/rpi/pigpio/) instead of Rpi.GPIO, so for consistency I would use it here also.

The interface is a bit clunky, bit it has more advanced features. Example use: https://github.com/SurrogateInc/surrortg-sdk/blob/main/games/claw/game.py

ollipal commented on games/ninswitch/game_simple_reset.py:
```diff
@@ -1,11 +1,32 @@
 from surrortg import Game
```
I'd keep the `game_simple.py` name. Creators can then optionally connect the reset GPIO pin.

ollipal commented on games/ninswitch/game_simple_reset.py:
```diff
@@ -44,6 +66,12 @@ async def on_init(self):
     """
 
+    async def on_prepare(self):
+        send_sig(nsg_reset, ON)
```
I hope that resetting the board between games is enough, so I would propose removing `reset_trinkets` class altogether.

Then I would wrap lines 70-74 to a method, and just call it here:
```python
async def on_prepare(self):
    await self.reset_trinket()
```

DaneHalle replied:
I tested it and what is proposed doesn't work directly. However this does work
```python
async def on_prepare(self):
        await ResetTrinkets().on()
        await ResetTrinkets().off()
        await asyncio.sleep(2)
```

ollipal reviewed (commented):
Both of the features are really nice additions! There are some minor suggestions to the code, what do you think?

ollipal commented:
You can for example run parts of the [pipelines](https://github.com/SurrogateInc/surrortg-sdk/blob/main/.github/workflows/main.yml) locally to figure out the current issues:

flake8:
```
pip install flake8==3.7.9
flake8 --config=.flake8
```

black:
```
pip install black==19.10b0
black --check --diff --config=pyproject.toml . 
```

Alternatively you can use pre-commit, the instructions are in the README.md.

On the git history there seems to be lots of unrelated files added and removed, you can get rid of them with [interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History). I can help you with that if that is something you haven't done before.

DaneHalle commented:
Oh wow looks like fixing flake8 issues (it was wanting new lines **before** the class definitions and not after) also fixed the black issue.  If you want to Feel free to comment on anything else and I will take a look at it or explain why I found it to work best that way.

ollipal commented on games/ninswitch/game_simple.py:
```diff
@@ -1,15 +1,36 @@
+OFF = 1
+
+
+class ResetTrinkets(Switch):
```
Based on our own games, it should be enough to reset trinkets between the games.

So I would replace the `ResetTrinkets` class with this function:
```python
async def reset_trinket():
    pi.write(nsg_reset, ON)
    logging.info(f"\t{seat} | TRINKET_RESET down")
    pi.write(nsg_reset, OFF)
    logging.info(f"\t...{seat} | TRINKET_RESET up")
    await asyncio.sleep(2)
```

and remove the input registered to it + modify on_prepare.

This simplification would make the `game_simple.py` a bit easier to understand for the creators.

DaneHalle replied:
In the first weekend running botw, there were probably 10+ times where someone in game had the trinket freeze up on them. If you feel that having input to allow users to manually reset should that happen isn't needed, I will go through with removing it.

ollipal commented on games/ninswitch/game_simple.py:
```diff
@@ -36,6 +57,7 @@ async def on_init(self):
                 "right_stick": NSSwitch(self.nsg, NSButton.RIGHT_STICK),
                 "home": NSSwitch(self.nsg, NSButton.HOME),
                 "capture": NSSwitch(self.nsg, NSButton.CAPTURE),
+                "reset_trinkets": ResetTrinkets(),
```
remove this line

ollipal commented on games/ninswitch/game_simple.py:
```diff
@@ -44,6 +66,11 @@ async def on_init(self):
     on_config, on_prepare, on_pre_game, on_countdown, on_start...
     """
 
+    async def on_prepare(self):
```
change this just to:

```python
    async def on_prepare(self):
        await reset_trinket()

```

ollipal commented on games/ninswitch/game_imagerec.py:
```diff
@@ -32,11 +34,31 @@
 ]
 
 
+class CaptureScreen(Switch):
```
I'm quite sure that you should never open two `AsyncVideoCapture.create("/dev/video21")`s at the same time, and this is why we should try to save frames with only one capture, even when there are two methods to do so.

Can you try once more with `global`? I think something like this should work (after renaming SAVE_FRAMES to SAVE_ALL_FRAMES to better differenciate these frame saving methods):
```python
save_frame = False

class CaptureScreen(Switch):
    async def on(self, seat=0):
        logging.info(f"\t{seat} | Capturing_Frames down")
        global save_frame
        save_frame = True

    async def off(self, seat=0):
        logging.info(f"\t{seat} | Capturing_Frames up")
        global save_frame
        save_frame = False

# later inside image_rec_main:

    if SAVE_ALL_FRAMES or save_frame:
        cv2.imwrite(f"{SAVE_DIR_PATH}/{i}.jpg", frame)
        logging.info(f"SAVED {i}.jpg")
```

ollipal commented on games/ninswitch/game_imagerec.py:
```diff
@@ -32,11 +34,31 @@
 class NinSwitchImageRecGame(Game):
     async def on_init(self):
         # init controls
+        self.cap = await AsyncVideoCapture.create("/dev/video21")
```
remove this, check comment above

ollipal reviewed (changes requested):
Nice that you got flake8 and black working! Still few comments, but after these I think we are ready to merge.

Commit:
ninswitch: imagerec alternate capture added, README updated

user can now determine when to save a frame in addition to constantly
saving frames while the script is running

ollipal commented on games/ninswitch/game_imagerec.py:
```diff
@@ -35,26 +34,24 @@
-    def __init__(self, cap):
-        Path(SAVE_DIR_PATH).mkdir(parents=True, exist_ok=True)
-        self.cap = cap
+    def __init__(self):
```
Add line `save_individual_frame = False` below `SAVE_ALL_FRAMES = False`, and remove the whole `def __init__` from this class.

Using lower case letters instead of `SAVE_INDIVIDUAL_FRAME` is more appropriate, because the value does not stay constant.

ollipal commented on games/ninswitch/game_simple.py:
```diff
@@ -15,14 +14,13 @@
-    async def off(self, seat=0):
-        pi.write(nsg_reset, OFF)
-        logging.info(f"\t...{seat} | TRINKET_RESET up")
+async def ResetTrinket():
```
function names in snake case: `ResetTrinket()` --> `reset_trinket()`

ollipal reviewed (changes requested):
Small styling changes

ollipal commented:
Before merging the pull request, I plan on cleaning the git history a bit with interactive rebase.

But you want to do it by yourself, here is a simple tutorial: https://www.youtube.com/watch?v=tukOm3Afd8s
Commands **r** (reword) and **f** (fixup) are required only. Also pushing must be done with `git push -f`as the git history will change.

By combining commits the history will be easier to understand: directories and files that are not used in the end will get removed (such as `FourPlayerTrinket/)`, and you can combine commits such as "flake8 fix(?)" or "removed unused import" to the correct commit as if they were done like that in the first place.

But I can do this for you if you are not familiar with these commands.

DaneHalle commented:
I am not the most familiar with rebasing using Git and attempted to but ran into some issues. I will let you do it as I do not want to break something.

ollipal commented:
In the end I simplified the changes into two commits, one for reset pin and other for the capture + readme. Thank you for your help!
